src/transformStateWithClones.js

```diff
@@ -7,7 +7,41 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  const versionsOfState = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        const { extraData } = action;
+
+        Object.assign(copyState, extraData);
+        break;
+
+      case 'removeProperties':
+        const { keysToRemove } = action;
+
+        removeProperties(copyState, keysToRemove);
+        break;
+
+      case 'clear':
+        copyState = {};
+        break;
+
+      default:
+        throw Error('unknown actions type');
```

Comment:
Let's add to the error line the `action.type` that went there

Reply:
Done